File: 19. Remove Nth Node From End of List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        # attach a dummy node
        node = ListNode(None)
        node.next = head
        head = node
        
        # two runners
        fptr = head.next
        while fptr and n:
            n -= 1
            fptr = fptr.next
            
        sptr = head
        while fptr:
            sptr = sptr.next
            fptr = fptr.next
            
        # remove the node
        sptr.next = sptr.next.next
        return head.next
        

File: 19. Remove Nth Node From End of List/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode sptr, fptr;
        fptr = head;
        fptr = new ListNode(0);
        fptr.next = head;
        head = fptr;
        while (n >= 0) {
            fptr = fptr.next;
            n--;
        }
        
        sptr = head;
        while (fptr != null) {
            fptr = fptr.next;
            sptr = sptr.next;
        }
        
        sptr.next = sptr.next.next;
        return head.next;
    }
}
